man/mptp.1
.\" -*- coding: utf-8 -*-
.\" ============================================================================
.TH mptp 1 "Sep 11, 2023" "mptp 0.2.5" "USER COMMANDS"
.\" ============================================================================
.SH NAME
mptp \(em single-locus species delimitation
.\" ============================================================================
.SH SYNOPSIS
.\" left justified, ragged right
.ad l
Maximum-likelihood species delimitation:
.RS
\fBmptp\fR \-\-ml (\-\-single | \-\-multi) \-\-tree_file \fInewickfile\fR 
\-\-output_file \fIoutputfile\fR [\fIoptions\fR]
.PP
.RE
Species delimitation with support values:
.RS
\fBmptp\fR \-\-mcmc \fIpositive integer\fR (\-\-single | \-\-multi)
(\-\-mcmc_startnull | \-\-mcmc_startrandom | \-\-mcmc_startml) \-\-mcmc_log
\fIpositive integer\fR \-\-tree_file \fInewickfile\fR \-\-output_file
\fIoutputfile\fR [\fIoptions\fR]
.PP
.RE
.\" left and right justified (default)
.ad b
.\" ============================================================================
.SH DESCRIPTION
Species is one of the fundamental units of comparison in virtually all
subfields of biology, from systematics to anatomy, development, ecology,
evolution, genetics and molecular biology. The aim of \fBmptp\fR is to offer
an open source tool to infer species boundaries on a a given phylogenetic tree
based on the Poisson Tree Process (PTP) and the Multiple Poisson Tree Process
(mPTP) models.
.PP
\fBmptp\fR offers two methods for inferring species delimitation. First, a
maximum-likelihood based method that uses a dynamic programming approach to
infer an ML estimate. Second, an mcmc approach for sampling the space of
possible delimitations providing the user with support values on the tree clades.
Both approaches are available in two flavours: the PTP and the mPTP model. The
PTP model is specified by using the \fIsingle\fR switch and the mPTP by using
\fImulti\fR.
.\" ============================================================================
.SS Input
The input for \fBmptp\fR is a newick file that contains one phylogenetic tree,
i.e., branches express the expected number of substitutions per alignment site.
.\" ============================================================================
.SS Options
\fBmptp\fR parses a large number of command-line options. For easier
navigation, options are grouped below by theme.
.PP
General options:
.RS
.TP 9
.B \-\-help
Display help text and exit.
.TP
.B \-\-version
Output version information and exit.
.TP
.B \-\-quiet
Supress all output to stdout except for warnings and fatal error messages.
.TP
.BI \-\-tree_file \0filename
Input newick file that contains a phylogenetic tree. Can be rooted or unrooted.
.TP
.BI \-\-output_file \0filename
Specifies the prefix used for generating output files. For maximum-likelihood
species delimitation two files will be created. First, \fIfilename\fR.txt that
contains the actual delimitation and \fIfilename\fR.svg that contains an SVG
figure of the computed delimitation. For mcmc analyses, a file
\fIfilename\fR.txt is created that contains the newick tree with supports
values.
.TP
.BI \-\-outgroup\~ "comma-separated list of taxa"
All computations for species delimitation are carried out on rooted trees. This
option is used only (and is required) In case an unrooted tree was specified
with the \-\-tree_file option. \fImptp\fR roots the unrooted tree by
splitting the branch leading to the most recent common ancestor (MRCA) of the
comma-separated list of taxa into two branches of equal size and introducing a
new node (the root of the new rooted tree) that connects these two branches.
.TP
.BI \-\-outgroup_crop
Crops taxa specified with the \-\-outgroup option from the the tree.
.TP
.BI \-\-min_br \0real
Any branch lengths in the input tree smaller or equal than \fIreal\fR are
excluded (ignored) from the computations. In addition, for mcmc analyses,
subtrees that exclusively consist of branch lengths smaller or equal to
\fIreal\fR are completely ignored from the proposals (support values for those
clades are set to 0). (default: 0.0001)
.TP
.BI \-\-precision\~ "positive integer"
Specifies the precision of the decimal part of floating point numbers on output
(default: 7)
.TP
.BI \-\-minbr_auto \0filename
Automatically detects the minimum branch length from the p-distances of the
FASTA file \fIfilename\fR.
.TP
.BI \-\-tree_show
Show an ASCII version of the processed input tree (i.e. after it is rooted by,
potentially cropping, the outgroup).
.RE
.PP
.\" ============================================================================
Maximum-likelihood estimations:
.PP
.RS
Estimating the maximum-likelihood delimitation is triggered by the switch
\-\-ml followed by \-\-single (the PTP model) or \-\-ml \-\-multi (the mPTP
model). Note that these two methods affect how options \-\-output_file behaves
and can be controlled using the \-\-min_br switch. Both methods require a
rooted phylogenetic tree, however an unrooted tree may be specified in
conjuction with the option \-\-outgroup. In this case, \fImptp\fR roots it at
that outgroup (see General options, \-\-outgroup for more info). Note that both
methods output an SVG depiction of the ML delimitation. See Visualization for
more information on adjusting and fine-tuning the SVG output.
.PP
Both methods ignore discard branch lengths of size smaller than the size
specified using the \-\-min_br option. The PTP model then attempts to find a
connected subgraph of the rooted tree that (a) contains the root, and (b) the
sum of likelihoods of fitting the edges of that subgraph in one exponential
distribution and the remaining  edges in another (exponential distribution) is
maximized. With likelihood we mean the sums of the probability density function
with the mean defined as the reciprocal of the average of edge lengths in the
particular distribution.
.PP
.TP 9
.B \-\-ml \-\-single
Triggers the algorithm for computing an ML estimate of the delimitation using
the PTP model.
.TP
.B \-\-ml \-\-multi
Triggers the algorithm for computing an ML estimate of the delimitation using
the mPTP model.
.TP
.B \-\-pvalue \0real
Only used with the PTP model (specified with \-\-single). Sets the p-value for
performing a likelihood ratio test. Note that, there is no likelihood ratio test
for the mPTP model this test is not done. (default: 0.001)
.RE
.PP
.\" ============================================================================
MCMC method:
.PP
.RS
The MCMC method is triggered with the \-\-mcmc switch combined with either
\-\-single (the PTP model) or \-\-multi (the mPTP model). 
.PP
Some more stuff to write
.PP
.TP 9
.B \-\-mcmc\~ "positive integer" \-\-single
Triggers the algorithm for computing support values by taking the specified
number of MCMC samples (delimitations) using the PTP model.
.TP
.B \-\-mcmc\~ "positive integer" \-\-multi
Triggers the algorithm for computing support values by taking the specified
number of MCMC samples (delimitations) using the mPTP model.
.TP
.B \-\-mcmc_sample\~ "positive integer"
Sample only every n-th MCMC step.
.TP
.B \-\-mcmc_log
Log the scores (log-likelihood) for each MCMC sample in a file and create an SVG
plot.
.TP
.B \-\-mcmc_burnin\~ "positive integer"
Ignore all MCMC samples generated before the specified step. (default: 1)
.TP
.B \-\-mcmc_runs\~ "positive integer"
Perform multiple MCMC runs. If more than 1 run is specified, mptp will generate
one seed for each run based on the provided seed using the \-\-seed switch.
Output files will be generated for each run (default: 1)
.TP
.B \-\-mcmc_credible \0real
Specify the probability (0.0 to 1.0) for which to generate the credible interval
i.e., the probability the true number of species will fall within the credible
interval given the observed data. (default: 0.95)
.TP
.B \-\-mcmc_startnull
Start MCMC sampling from the null-model.
.TP
.B \-\-mcmc_startrandom
Start MCMC sampling from a random delimitation. 
.TP
.B \-\-mcmc_startrandom
Start MCMC sampling from the ML delimitation.
.TP
.B \-\-seed\~ "positive integer"
Specifies the seed for the pseudo-random number generator. (default: randomly
generated based on system time)
.RE
.PP
.\" ============================================================================
SVG Output:
.PP
.RS
The ML method generates one SVG file that visualizes the processed input tree
(i.e. after it is rooted by, potentially cropping, the outgroup) and marks the
subtrees corresponding to coalescent processes (the detected species groups)
with red color, while the speciation process is colored green.
.PP
The MCMC method generates one SVG file per run visualizing the processed
tree, and indicates the support value for each node, i.e., the percentage of
MCMC samples (delimitations) in which the particular node was part of the
speciation process.  A value of 1 means it was always in the speciation process
while a value of 0 means it was always in a coalescent process. The tree
branches are colored according to the support values of descendant nodes; a
support of value of 0 is colored with red, 1 with black, and values in between
are gradients of the two colors. Only support values above 0.5 are shown to
avoid packed numbers in dense branching events. In addition, if \-\-mcmc_log is
specified, an additional SVG image of log-likelihoods plots for each sampled
delimitation is created.
.PP
.TP 9
.B \-\-svg_width\~ "positive integer"
Sets the total width (including margins) of the SVG in pixels. (default: 1920)
.TP
.B \-\-svg_fontsize\~ "positive integer"
Size of font in SVG image. (default: 12)
.TP
.B \-\-svg_tipspacing\~ "positive integer"
Vertical space in pixels between taxa in SVG tree. (default: 20)
.TP
.B \-\-svg_legend_ratio \0real
Ratio (value between 0.0 and 1.0) of total tree length to be displayed as
legend line.  (default: 0.1)
.TP
.B \-\-svg_nolengend
Hide legend.
.TP
.B \-\-svg_marginleft\~ "positive integer"
Left margin in pixels. (default: 20)
.TP
.B \-\-svg_marginright\~ "positive integer"
Right margin in pixels. (default: 20)
.TP
.B \-\-svg_margintop\~ "positive integer"
Top margin in pixels. (default: 20)
.TP
.B \-\-svg_marginbottom\~ "positive integer"
Top margin in pixels. (default: 20)
.TP
.B \-\-svg_inner_radius\~ "positive integer"
Radius of inner nodes in pixels. (default: 0)
.RE
.PP
.\" ============================================================================
.SH EXAMPLES
.PP
Compute the maximum likelihood estimate using the mPTP model by discarding all
branches with length below or equal to 0.0001 
.PP
.RS
\fBmptp\fR \-\-ml \-\-multi \-\-min_br 0.0001 \-\-tree_file \fInewick.txt\fR
\-\-output_file \fIout\fR
.RE
.PP
Run an MCMC analysis of 100 million steps with the mPTP model, that logs every
one million-th step, ignores the first 2 million steps and discards all branches
with lengths smaller or equal to 0.0001. Use 777 as seed. The chain will start
from the ML delimitation (default).
.PP
.RS
\fBmptp\fR \-\-mcmc 100000000 \-\-multi \-\-min_br 0.0001 \-\-tree_file
\fInewick.txt\fR \-\-output_file \fIout\fR \-\-mcmc_log 1000000 \-\-mcmc_burnin
2000000 -seed 777
.RE
.PP
Perform an MCMC analysis of 5 runs, each of 100 million steps with the mPTP
model, log every one million-th step, ignore the first 2 million steps, and
detect the minimum branch length by specifying the FASTA file alignment.fa that
contains the alignment. Use 777 as seed. Start each run from a random
delimitation.
.PP
.RS
\fBmptp\fR \-\-mcmc 100000000 \-\-multi -\-\-mcmc_runs 5 \-\-mcmc_log 1000000
\-\-minbr_auto \fIalignment.fa\fR \-\-tree_file \fInewick.txt\fR
\-\-output_file \fIout\fR \-\-mcmc_burnin 2000000 -seed 777
\-\-mcmc_startrandom
.RE
.PP
.\"
.\" ============================================================================
.SH AUTHORS
Implementation by Tomas Flouri, Sarah Lutteropp and Paschalia Kapli. Additional
PTP and mPTP model authors include Kassian Kobert, Jiajie Zhang, Pavlos
Pavlidis, and Alexandros Stamatakis.
.SH REPORTING BUGS
Submit suggestions and bug-reports at
<https://github.com/Pas-Kapli/mptp/issues>, or e-mail Tomas Flouri
<Tomas.Flouri@h-its.org>.
.\" ============================================================================
.SH AVAILABILITY
Source code and binaries are available at
<https://github.com/Pas-Kapli/mptp>.
.\" ============================================================================
.SH COPYRIGHT
Copyright (C) 2015-2017, Tomas Flouri, Sarah Lutteropp, Paschalia Kapli
.PP
All rights reserved.
.PP
Contact: Tomas Flouri <Tomas.Flouri@h-its.org>,
Scientific Computing, Heidelberg Insititute for Theoretical Studies,
69118 Heidelberg, Germany
.PP
This software is licensed under the terms of the GNU Affero General Public
License version 3.
.PP
\fBGNU Affero General Public License version 3\fR
.PP
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU Affero General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.
.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
details.
.PP
You should have received a copy of the GNU Affero General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.
.SH VERSION HISTORY
New features and important modifications of \fBmptp\fR (short lived or minor
bug releases may not be mentioned):
.RS
.TP
.BR v0.1.0\~ "released June 27th, 2016"
First public release.
.TP
.BR v0.1.1\~ "released July 15th, 2016"
Bug fix (now LRT test is not printed in output file when using --multi)
.TP
.BR v.0.2.0\~ "released September 27th, 2016"
Fixed floating point exception error when constructing random trees, caused
from dividing by zero.  Changed allocation from malloc to calloc, as it caused
unititialized variables when converting unrooted trees to rooted when using the
MCMC method. Fixed sample size for the AIC with a correction for finite sample
sizes.
.TP
.BR v.0.2.1\~ "released October 18th, 2016"
Updated ASV to consider only coalescent roots of ML delimitation. Removed
assertion stopping mptp when using random starting delimitations for the MCMC
method.
.TP
.BR v0.2.2\~ "released January 31st, 2017"
Fixed regular expressions to allow scientific notation for branch lengths when
parsing trees.  Improved the accuracy of ASV score by also taking into account
tips forming coalescent roots.  Fixed memory leaks that occur when parsing
incorrectly formatted trees.
.TP
.BR v0.2.3\~ "released July 25th, 2017"
Replaced hsearch() with custom hashtable. Fixed minor output error messages.
.TP
.BR v0.2.4\~ "released May 14th, 2018"
If we do not manage to generate a random starting delimitation with the wanted
number of species (randomly chosen), we use the currently generated
delimitation instead.
.TP
.BR v0.2.5\~ "released Sep 9th, 2023"
Added likelihood ratio test for the multi method. Added implementation for the
incomplete gamma function, and removed dependency for GNU scientific library.
.RE
.LP


ChangeLog.md
# Change Log
All notable changes to `mptp` will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.5] - 2023-09-11
### Added
 - Likelihood ratio test for the multi method
 - Added implementation for the incomplete gamma function
### Removed
 - Dependency for GNU scientific library

## [0.2.4] - 2018-05-14
### Fixed
 - If we do not manage to generate a random starting delimitation with the
   wanted number of species (randomly chosen), we use the currently generated
   delimitation instead.

## [0.2.3] - 2017-07-25
### Fixed
 - Replaced hsearch which was causing problems on APPLE with custom hashtable
 - Corrected file name in error messages when failing to open files

## [0.2.2] - 2017-01-31
### Fixed
 - Regular expressions now allow scientific notation when parsing branch lengths
 - Improved accuracy of ASV score (takes into account tip species)
 - Memory leaks when parsing incorrectly formatted trees

## [0.2.1] - 2016-10-18
### Fixed
 - Updated ASV to consider only coalescent roots of ML delimitation
 - Assertion stopping mptp when using random starting delimitations for MCMC

## [0.2.0] - 2016-09-27
### Fixed
 - Floating point exception error when constructing random trees caused from
   division by zero
 - Allocation with malloc caused uninitialized variables when converting unrooted
   tree to rooted for the MCMC method
 - Sample size for the the AIC with a correction for finite sample sizes


src/mptp.c
/*
    Copyright (C) 2015 Tomas Flouri, Sarah Lutteropp

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact: Tomas Flouri <Tomas.Flouri@h-its.org>,
    Heidelberg Institute for Theoretical Studies,
    Schloss-Wolfsbrunnenweg 35, D-69118 Heidelberg, Germany
*/

#include "mptp.h"

static char * progname;
static char progheader[80];
char * cmdline;

/* global error message buffer */
char errmsg[200] = {0};

/* global pseudo-random number generator 48-bit state */
unsigned short global_xsubi[3];

/* number of mandatory options for the user to input */
static const char mandatory_options_count = 2;
static const char * mandatory_options_list = " --tree_file --output_file";

/* options */
int pll_errno;
int opt_quiet;
int opt_precision;
int opt_svg_showlegend;
long opt_help;
long opt_version;
long opt_treeshow;
long opt_method;
long opt_mcmc_sample;
long opt_mcmc_steps;
long opt_mcmc_log;
long opt_mcmc_startnull;
long opt_mcmc_startrandom;
long opt_mcmc_startml;
long opt_mcmc_burnin;
long opt_mcmc_runs;
long opt_seed;
long opt_mcmc;
long opt_ml;
long opt_multi;
long opt_single;
long opt_crop;
long opt_svg;
long opt_svg_width;
long opt_svg_fontsize;
long opt_svg_tipspace;
long opt_svg_marginleft;
long opt_svg_marginright;
long opt_svg_margintop;
long opt_svg_marginbottom;
long opt_svg_inner_radius;
double opt_mcmc_credible;
double opt_svg_legend_ratio;
double opt_pvalue;
double opt_minbr;
char * opt_treefile;
char * opt_outfile;
char * opt_outgroup;
char * opt_pdist_file;

static struct option long_options[] =
{
  {"help",               no_argument,       0, 0 },  /*  0 */
  {"version",            no_argument,       0, 0 },  /*  1 */
  {"quiet",              no_argument,       0, 0 },  /*  2 */
  {"tree_file",          required_argument, 0, 0 },  /*  3 */
  {"tree_show",          no_argument,       0, 0 },  /*  4 */
  {"output_file",        required_argument, 0, 0 },  /*  5 */
  {"outgroup",           required_argument, 0, 0 },  /*  6 */
  {"pvalue",             required_argument, 0, 0 },  /*  7 */
  {"minbr",              required_argument, 0, 0 },  /*  8 */
  {"svg_width",          required_argument, 0, 0 },  /*  9 */
  {"svg_fontsize",       required_argument, 0, 0 },  /* 10 */
  {"svg_tipspacing",     required_argument, 0, 0 },  /* 11 */
  {"svg_legend_ratio",   required_argument, 0, 0 },  /* 12 */
  {"svg_nolegend",       no_argument,       0, 0 },  /* 13 */
  {"svg_marginleft",     required_argument, 0, 0 },  /* 14 */
  {"svg_marginright",    required_argument, 0, 0 },  /* 15 */
  {"svg_margintop",      required_argument, 0, 0 },  /* 16 */
  {"svg_marginbottom",   required_argument, 0, 0 },  /* 17 */
  {"svg_inner_radius",   required_argument, 0, 0 },  /* 18 */
  {"precision",          required_argument, 0, 0 },  /* 19 */
  {"mcmc_sample",        required_argument, 0, 0 },  /* 20 */
  {"mcmc_log",           no_argument,       0, 0 },  /* 21 */
  {"seed",               required_argument, 0, 0 },  /* 22 */
  {"mcmc_startnull",     no_argument,       0, 0 },  /* 23 */
  {"mcmc_burnin",        required_argument, 0, 0 },  /* 24 */
  {"mcmc_startrandom",   no_argument,       0, 0 },  /* 25 */
  {"mcmc_runs",          required_argument, 0, 0 },  /* 26 */
  {"minbr_auto",         required_argument, 0, 0 },  /* 27 */
  {"outgroup_crop",      no_argument,       0, 0 },  /* 28 */
  {"mcmc_credible",      required_argument, 0, 0 },  /* 29 */
  {"mcmc",               required_argument, 0, 0 },  /* 30 */
  {"ml",                 no_argument,       0, 0 },  /* 31 */
  {"single",             no_argument,       0, 0 },  /* 32 */
  {"multi",              no_argument,       0, 0 },  /* 33 */
  {"mcmc_startml",       no_argument,       0, 0 },  /* 34 */
  { 0, 0, 0, 0 }
};

void args_init(int argc, char ** argv)
{
  int option_index = 0;
  int c;
  int mand_options = 0;

  /* set defaults */

  progname = argv[0];

  opt_help = 0;
  opt_version = 0;
  opt_treeshow = 0;
  opt_treefile = NULL;
  opt_outfile = NULL;
  opt_outgroup = NULL;
  opt_pdist_file = NULL;
  opt_quiet = 0;
  opt_pvalue = 0.001;
  opt_minbr = 0.0001;
  opt_precision = 7;
  opt_mcmc_steps = 0;
  opt_mcmc_sample = 1000;
  opt_mcmc_startnull = 0;
  opt_mcmc_startrandom = 0;
  opt_mcmc_startml = 0;
  opt_mcmc_log = 0;
  opt_mcmc_burnin = 1;
  opt_mcmc_runs = 1;
  opt_mcmc_credible = 0.95;
  opt_seed = (long)time(NULL);
  opt_crop = 0;
  opt_ml = 0;
  opt_mcmc = 0;
  opt_method = PTP_METHOD_MULTI;
  opt_multi = 0;
  opt_single = 0;

  opt_svg_width = 1920;
  opt_svg_fontsize = 12;
  opt_svg_tipspace = 20;
  opt_svg_legend_ratio = 0.1;
  opt_svg_showlegend = 1;
  opt_svg_marginleft = 20;
  opt_svg_marginright = 20;
  opt_svg_margintop = 20;
  opt_svg_marginbottom = 20;
  opt_svg_inner_radius = 0;

  while ((c = getopt_long_only(argc, argv, "", long_options, &option_index)) == 0)
  {
    char * end;
    switch (option_index)
    {
      case 0:
        opt_help = 1;
        break;

      case 1:
        opt_version = 1;
        break;

      case 2:
        opt_quiet = 1;
        break;

      case 3:
        free(opt_treefile);
        opt_treefile = optarg;
        break;

      case 4:
        opt_treeshow = 1;
        break;

      case 5:
        opt_outfile = optarg;
        break;

      case 6:
        opt_outgroup = optarg;
        break;

      case 7:
        opt_pvalue = strtod(optarg, &end);
        if (end == optarg) {
          fatal(" is not a valid number.\n");
        }
        break;

      case 8:
        opt_minbr = strtod(optarg, &end);
        if (end == optarg) {
          fatal(" is not a valid number.\n");
        }
        break;

      case 9:
        opt_svg_width = atoi(optarg);
        break;

      case 10:
        opt_svg_fontsize = atol(optarg);
        break;

      case 11:
        opt_svg_tipspace = atol(optarg);
        break;

      case 12:
        opt_svg_legend_ratio = atof(optarg);
        break;

      case 13:
        opt_svg_showlegend = 0;
        break;

      case 14:
        opt_svg_marginleft = atol(optarg);
        break;

      case 15:
        opt_svg_marginright = atol(optarg);
        break;

      case 16:
        opt_svg_margintop = atol(optarg);
        break;

      case 17:
        opt_svg_marginbottom = atol(optarg);
        break;

      case 18:
        opt_svg_inner_radius = atol(optarg);
        break;

      case 19:
        opt_precision = atoi(optarg);
        break;

      case 20:
        opt_mcmc_sample = atol(optarg);
        break;

      case 21:
        opt_mcmc_log = 1;
        break;

      case 22:
        opt_seed = atol(optarg);
        break;

      case 23:
        opt_mcmc_startnull = 1;
        break;

      case 24:
        opt_mcmc_burnin = atol(optarg);
        break;

      case 25:
        opt_mcmc_startrandom = 1;
        break;

      case 26:
        opt_mcmc_runs = atol(optarg);
        break;

      case 27:
        free(opt_pdist_file);
        opt_pdist_file = optarg;
        break;

      case 28:
        opt_crop = 1;
        break;

      case 29:
        opt_mcmc_credible = atof(optarg);
        break;

      case 30:
        opt_mcmc = 1;
        opt_mcmc_steps = atol(optarg);
        break;

      case 31:
        opt_ml = 1;
        break;

      case 32:
        opt_method = PTP_METHOD_SINGLE;
        opt_single = 1;
        break;

      case 33:
        opt_method = PTP_METHOD_MULTI;
        opt_multi = 1;
        break;

      case 34:
        opt_mcmc_startml = 1;
        break;

      default:
        fatal("Internal error in option parsing");
    }
  }

  if (c != -1)
    exit(EXIT_FAILURE);

  int commands  = 0;

  /* check for mandatory options */
  if (opt_treefile)
    mand_options++;
  if (opt_outfile)
    mand_options++;

  /* check for number of independent commands selected */
  if (opt_version)
    commands++;
  if (opt_help)
    commands++;
  if (opt_pdist_file)
    commands++;
  if (opt_mcmc)
    commands++;
  if (opt_ml)
    commands++;

  /* if more than one independent command, fail */
  if (commands > 1)
  {
    if (!(commands == 2 && opt_pdist_file && (opt_ml || opt_mcmc)))
      fatal("More than one command specified");
  }

  /* if more than one independent command, fail */
  if (opt_mcmc_startrandom + opt_mcmc_startnull + opt_mcmc_startml > 1)
    fatal("You can only select one out of --mcmc_startrandom, --mcmc_startnull, --mcmc_startml");

  /* if more than one independent command, fail */
  if (opt_multi && opt_single)
    fatal("You can either specify --multi or --single, but not both at once.");

  /* if no command specified, turn on --help */
  if (!commands)
  {
    opt_help = 1;
    return;
  }
  /* check for mandatory options */
  if (!opt_version && !opt_help)
    if (mand_options != mandatory_options_count)
      fatal("Mandatory options are:\n\n%s", mandatory_options_list);

}

void cmd_help()
{
  fprintf(stderr,
          "Usage: %s [OPTIONS]\n", progname);
  fprintf(stderr,
          "\n"
          "Examples:\n"
          "  mptp --ml --multi --tree_file tree.newick --output_file output\n"
          "  mptp --mcmc 50000000 --multi --mcmc_sample 1000000 --mcmc_burnin 1000000 --tree_file tree.newick --output_file output\n\n"
          "General options:\n"
          "  --help                    display help information.\n"
          "  --version                 display version information.\n"
          "  --tree_show               display an ASCII version of the tree.\n"
          "  --multi                   Use one lambda per coalescent (this is default).\n"
          "  --single                  Use one lambda for all coalescent.\n"
          "  --ml                      Maximum-likelihood heuristic.\n"
          "  --mcmc INT                Support values for the delimitation (INT steps).\n"
          "  --mcmc_sample INT         Sample every INT iteration (default: 1000).\n"
          "  --mcmc_log                Log samples and create SVG plot of log-likelihoods.\n"
          "  --mcmc_burnin INT         Ignore all MCMC steps below threshold.\n"
          "  --mcmc_runs INT           Perform multiple MCMC runs.\n"
          "  --mcmc_credible <0..1>    Credible interval (default: 0.95).\n"
          "  --mcmc_startnull          Start each run with the null model (one single species).\n"
          "  --mcmc_startrandom        Start each run with a random delimitation.\n"
          "  --mcmc_startml            Start each run with the delimitation obtained by the Maximum-likelihood heuristic.\n"
          "  --pvalue REAL             Set p-value for LRT (default: 0.001)\n"
          "  --minbr REAL              Set minimum branch length (default: 0.0001)\n"
          "  --minbr_auto FILENAME     Detect minimum branch length from FASTA p-distances\n"
          "  --outgroup TAXA           Root unrooted tree at outgroup (default: taxon with longest branch).\n"
          "  --outgroup_crop           Crop outgroup from tree\n"
          "  --quiet                   only output warnings and fatal errors to stderr.\n"
          "  --precision INT           Precision of floating point numbers on output (default: 7).\n"
          "  --seed                    Seed for pseudo-random number generator.\n"
          "\n"
          "Input and output options:\n"
          "  --tree_file FILENAME      tree file in newick format.\n"
          "  --output_file FILENAME    output file name.\n"
          "\n"
          "Visualization options:\n"
          "  --svg_width INT           Width of SVG tree in pixels (default: 1920).\n"
          "  --svg_fontsize INT        Size of font in SVG image. (default: 12)\n"
          "  --svg_tipspacing INT      Vertical space between taxa in SVG tree (default: 20).\n"
          "  --svg_legend_ratio <0..1> Ratio of total tree length to be displayed as legend line.\n"
          "  --svg_nolegend            Hides legend.\n"
          "  --svg_marginleft INT      Left margin in pixels (default: 20).\n"
          "  --svg_marginright INT     Right margin in pixels (default: 20).\n"
          "  --svg_margintop INT       Top margin in pixels (default: 20).\n"
          "  --svg_marginbottom  INT   Bottom margin in pixels (default: 20).\n"
          "  --svg_inner_radius INT    Radius of inner nodes in pixels (default: 0).\n"
         );
}

static rtree_t * load_tree(void)
{
  /* parse tree */
  if (!opt_quiet)
    fprintf(stdout, "Parsing tree file...\n");

  rtree_t * rtree = rtree_parse_newick(opt_treefile);

  if (!rtree)
  {
    unsigned int tip_count;
    utree_t * utree = utree_parse_newick(opt_treefile, &tip_count);
    if (!utree)
      fatal("Tree is neither unrooted nor rooted.");

    if (!opt_quiet)
    {
      fprintf(stdout, "Loaded unrooted tree...\n");
      fprintf(stdout, "Converting to rooted tree...\n");
    }

    /* if outgroup was not specified, get the node with the longest branch */
    utree_t * og_root = NULL;

    /* if outgroup was not specified, get the tip with the longest branch */
    if (!opt_outgroup)
    {
      og_root = utree_longest_branchtip(utree, tip_count);
      assert(og_root);
      fprintf(stdout,
              "Selected %s as outgroup based on longest tip-branch criterion\n",
              og_root->label);
    }
    else
    {
      /* get LCA of out group */
      og_root = utree_outgroup_lca(utree, tip_count);
      if (!og_root)
      {
        utree_destroy(utree);
        fatal("Outgroup must be a single tip or a list of all tips of a subtree");
      }
    }

    if (opt_crop)
    {
      rtree = utree_crop(og_root);
    }
    else
    {
      rtree = utree_convert_rtree(og_root);
    }

    utree_destroy(utree);
  }
  else
  {
    if (!opt_quiet)
      fprintf(stdout, "Loaded rooted tree...\n");
      
    if (opt_crop)
    {
      if (!opt_outgroup)
        fatal("--outgroup must be specified when using --outgroup_crop.");

      /* get LCA of outgroup */
      rtree_t * og_root = get_outgroup_lca(rtree);

      /* crop outgroup from tree */
      rtree = rtree_crop(rtree,og_root);
      if (!rtree)
        fatal("Cropping the outgroup leads to less than two tips.");
    }
  }

  return rtree;
}

void cmd_auto()
{
  rtree_t * rtree = load_tree();

  detect_min_bl(rtree, 0);

  /* deallocate tree structure */
  rtree_destroy(rtree);
}

void cmd_ml(void)
{
  rtree_t * rtree = load_tree();
  if (opt_pdist_file)
    opt_minbr = detect_min_bl(rtree, 1);

  dp_init(rtree);
  dp_set_pernode_spec_edges(rtree);
  dp_ptp(rtree, opt_method);
  dp_free(rtree);

  if (opt_treeshow)
    rtree_show_ascii(rtree);

  cmd_svg(rtree, opt_seed, "svg");

  /* deallocate tree structure */
  rtree_destroy(rtree);

  if (!opt_quiet)
    fprintf(stdout, "Done...\n");
}

void cmd_multirun(void)
{
  if (opt_mcmc_steps == 0)
    fatal("The number of runs specified after --mcmc must be a positive integer greater than zero");

  if (opt_mcmc_burnin < 1 || opt_mcmc_burnin > opt_mcmc_steps)
    fatal("--opt_mcmc_burnin must be a positive integer smaller or equal to --opt_mcmc_steps");

  if (opt_mcmc_credible < 0 || opt_mcmc_credible > 1)
    fatal("--opt_mcmc_credible must be a real number between 0 and 1");

  rtree_t * rtree = load_tree();

  if (opt_pdist_file)
    opt_minbr = detect_min_bl(rtree, 1);

  multirun(rtree, opt_method);

  if (opt_treeshow)
    rtree_show_ascii(rtree);

  if (!opt_quiet)
    fprintf(stdout, "Done...\n");

}

void getentirecommandline(int argc, char * argv[])
{
  int len = 0;
  int i;

  for (i = 0; i < argc; ++i)
    len += strlen(argv[i]);

  cmdline = (char *)xmalloc((size_t)(len + argc + 1));
  cmdline[0] = 0;

  for (i = 0; i < argc; ++i)
  {
    strcat(cmdline, argv[i]);
    strcat(cmdline, " ");
  }
}

void fillheader()
{
  snprintf(progheader, 80,
           "%s %s_%s, %1.fGB RAM, %ld cores",
           PROG_NAME, PROG_VERSION, PROG_ARCH,
           arch_get_memtotal() / 1024.0 / 1024.0 / 1024.0,
           arch_get_cores());
}

void show_header()
{
  fprintf(stdout, "%s\n", progheader);
  fprintf(stdout, "https://github.com/Pas-Kapli/mptp\n");
  fprintf(stdout,"\n");
}

int main (int argc, char * argv[])
{
  fillheader();
  getentirecommandline(argc, argv);

  args_init(argc, argv);

  show_header();

  /* init random number generator and maintain compatibility with srand48 */
  random_init(global_xsubi,opt_seed);

  if (opt_help)
  {
    cmd_help();
  }
  else if (opt_pdist_file && !(opt_mcmc || opt_ml))
  {
    cmd_auto();
  }
  else if (opt_mcmc)
  {
    cmd_multirun();
  }
  else if (opt_ml)
  {
    cmd_ml();
  }

  free(cmdline);
  return (0);
}
